### PR TITLE
Code Cleaning, Multiple Issues Resolved

### PR DESCRIPTION
**Pull request by HealthyDoggo**
Changing verbose_name and verbose_name_plural to ensure that Media Gallery ensures so in the admin and not as Media gallery, as stated in #81  Also, resolved issue #82 by changing .CASCADE to .DO_NOTHING for elements in the navbar model. Finally, resolved a typo in folder structure where util was spelt utill, and changed in all files to ensure that everything didn't break.

### CODE REVIEW

**Comment by HealthyDoggo** [comment]
Added to this commit input validation as stated in issue #59 including a custom error for invalid hex codes.

**Comment by HealthyDoggo** [comment]
Now pushing a fix for issue #83 replacing missing navbar or media_gallery elements on the Theme export method with empty strings to ensure that the Theme export method has the same support for missing elements as the Theme itself.

File: django_dynamic_theme/models.py [hunk 1]
```diff
@@ -190,9 +193,15 @@ def export(self) -> str:
         """
         Exports all listed configurations as string in SCSS format.
         """
-        return f"""body {{{self.background.export()}}}
-{self.media_gallery.export()}
-{self.navbar.export()}"""
+        background = self.background.export()
+        try: mediagallery = self.media_gallery.export()
```
Comment (Segelzwerg) [inline django_dynamic_theme/models.py:197]: Can you check for None instead of try?

File: django_dynamic_theme/utill/__init__.py
Comment (Segelzwerg) [inline django_dynamic_theme/utill/__init__.py:1]: We still need an __init__.py file

File: django_dynamic_theme/utill/color_converter.py
Comment (Segelzwerg) [inline django_dynamic_theme/utill/color_converter.py:1]: This file is also needed

File: django_dynamic_theme/utill/scss_editor.py
Comment (Segelzwerg) [inline django_dynamic_theme/utill/scss_editor.py:1]: We need this file, too.

File: django_dynamic_theme/errors.py [hunk 2]
```diff
@@ -5,3 +5,7 @@ class ThemeMissingError(Exception):
     """
     Is raised if not theme is set.
     """
+class NonHexValueError(Exception):
```
Comment (Segelzwerg) [inline django_dynamic_theme/errors.py:8]: Where do we use it?

**Review by Segelzwerg — CHANGES_REQUESTED** [review]
Great that you want to contribute. I think it needs a few fixes. I left a few comments on the code.

File: django_dynamic_theme/utill/scss_editor.py
Comment (HealthyDoggo) [reply django_dynamic_theme/utill/scss_editor.py:1]: oddly enough, i didn’t delete these files in my local codebase. Have no idea why github has assumed i have deleted them.

Comment (HealthyDoggo) [reply django_dynamic_theme/utill/scss_editor.py:1]: think it was something to do with me changing the folder name of utils - i’ll look into it this evening

File: django_dynamic_theme/models.py (hunk 1 above)
Comment (HealthyDoggo) [reply django_dynamic_theme/models.py:197]: yes will do this evening, apologies for that

File: django_dynamic_theme/utill/scss_editor.py
Comment (Segelzwerg) [reply django_dynamic_theme/utill/scss_editor.py:1]: Yes probably that is the issue. No worries. Take your time.

**Comment by HealthyDoggo** [comment]
Just pushed a fix to the previous suggestions - ensured that new util folder was git added (facepalm..) and that I was checking for None when exporting themes with empty elements instead of just try, excepting.

File: django_dynamic_theme/errors.py (hunk 2 above)
Comment (HealthyDoggo) [reply django_dynamic_theme/errors.py:8]: this was used in the deleted utils folders, use is present and visible now with new commit.

**Review by Segelzwerg — CHANGES_REQUESTED** [review]
Almost there. Can you look at the hex checking function?

**Comment by HealthyDoggo** [comment]
> Almost there. Can you look at the hex checking function?

Sorry slightly unclear here, this check is simply for code formatting correct? I just need to address whitespaces, line lengths, etc.?

**Comment by Segelzwerg** [comment]
> > Almost there. Can you look at the hex checking function?
> 
> Sorry slightly unclear here, this check is simply for code formatting correct? I just need to address whitespaces, line lengths, etc.?

Yes that's right. On the code itself I posted a link that might help reduce the complexity.

**Comment by HealthyDoggo** [comment]
Now pushed a new branch that hopefully fixes all PEP8 infractions.

**Comment by Segelzwerg** [comment]
@HealthyDoggo  I'm not sure if the changes require a migration? Could verify by running `python make_migrations.py`

**Comment by HealthyDoggo** [comment]
> @HealthyDoggo I'm not sure if the changes require a migration? Could verify by running `python make_migrations.py`

yep, just ran make_migrations and both migrations for meta options on mediagallery and the navbar field on theme have been made. Do I need to run migrate, as you would with normal Django, or is this fine?

File: django_dynamic_theme/util/color_converter.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        return False
+
+    for i in range(1, len(str)):
+        condition1 = (str[i] >= '0' and str[i] <= '9')
+        condition2 = (str[i] >= 'a' and str[i] <= 'f')
+        condition3 = (str[i] >= 'A' or str[i] <= 'F')
+        if (not ((condition1) or (condition2) or (condition3))):
+            return False
+
+    return True
```
[the remark this answers is not in the document]
Comment (HealthyDoggo) [reply django_dynamic_theme/util/color_converter.py]: Yep, added test_input_validation that checks for a NonHexValueError with an invalid RGB Hex (wrong length).

File: tests/test_utils.py [hunk 4]
```diff
@@ -27,3 +28,7 @@ def test_hex_to_rgb_converter(self):
         hex = "#1234FF"
         expected_tuple = (18, 52, 255)
         self.assertEqual(expected_tuple, hex_to_rgb_tuple(hex))
+
```
Comment (Segelzwerg) [inline tests/test_utils.py:31]:
```suggestion

```

**Review by Segelzwerg — CHANGES_REQUESTED** [review]

**Comment by HealthyDoggo** [comment]
Sorry about the continued changes - first time conforming to PEP8 guidelines so simply not used to adjusting my code to conform. This is also one of my first times contributing to a GitHub project, despite programming in Python for around 5 years. Thanks for your continued patience!

**Comment by Segelzwerg** [comment]
> Sorry about the continued changes - first time conforming to PEP8 guidelines so simply not used to adjusting my code to conform. This is also one of my first times contributing to a GitHub project, despite programming in Python for around 5 years. Thanks for your continued patience!

All good, no worries. Thanks for sticking to it and being patient with the linter.

**Comment by Segelzwerg** [comment]
Sorry, closed by mistake :D

File: tests/test_utils.py [hunk 5]
```diff
@@ -27,3 +28,6 @@ def test_hex_to_rgb_converter(self):
         hex = "#1234FF"
         expected_tuple = (18, 52, 255)
         self.assertEqual(expected_tuple, hex_to_rgb_tuple(hex))
+    def test_input_validation(self):
```
Comment (Segelzwerg) [inline tests/test_utils.py:31]:
```suggestion
        self.assertEqual(expected_tuple, hex_to_rgb_tuple(hex))
    
    def test_input_validation(self):
```

**Review by Segelzwerg — CHANGES_REQUESTED** [review]

File: tests/test_utils.py [hunk 6]
```diff
@@ -27,3 +28,7 @@ def test_hex_to_rgb_converter(self):
         hex = "#1234FF"
         expected_tuple = (18, 52, 255)
         self.assertEqual(expected_tuple, hex_to_rgb_tuple(hex))
+    
+    def test_input_validation(self):
```
Comment (Segelzwerg) [inline tests/test_utils.py:32]:
```suggestion
        self.assertEqual(expected_tuple, hex_to_rgb_tuple(hex))

    def test_input_validation(self):
```

Comment (Segelzwerg) [inline tests/test_utils.py:32]:
```suggestion
        self.assertEqual(expected_tuple, hex_to_rgb_tuple(hex))

    def test_input_validation(self):
```

File: django_dynamic_theme/models.py [hunk 7]
```diff
@@ -102,6 +102,10 @@ class MediaGallery(ThemeElement):
     item_align = models.CharField(max_length=50)
     row_margin_top = models.CharField(max_length=50)
 
+    class Meta:
+        verbose_name = "Media Gallery"
```
Comment (Segelzwerg) [inline django_dynamic_theme/models.py:106]:
```suggestion
    class Meta:
    """Meta definitions of Media Gallery"""
        verbose_name = "Media Gallery"
```

File: django_dynamic_theme/models.py [hunk 8]
```diff
@@ -102,6 +102,10 @@ class MediaGallery(ThemeElement):
     item_align = models.CharField(max_length=50)
     row_margin_top = models.CharField(max_length=50)
 
+    class Meta:
```
Comment (Segelzwerg) [inline django_dynamic_theme/models.py:105]:
```suggestion
    row_margin_top = models.CharField(max_length=50)

    # pylint: disable=too-few-public-methods
    class Meta:
```

File: django_dynamic_theme/util/color_converter.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
```
Comment (Segelzwerg) [inline django_dynamic_theme/util/color_converter.py:40]:
```suggestion
def is_valid_hexa_code(str):
    """
    Checks if the a string is a hex value.
    :param value: the string to be checked.
    :returns: True if hex value else False.
    """
    
    if value[0] != '#':
        return False

    if not (len(value) == 4 or len(value) == 7):
        return False

    for i in range(1, len(str)):
        condition1 = (str[i] >= '0' and str[i] <= '9')
        condition2 = (str[i] >= 'a' and str[i] <= 'f')
        condition3 = (str[i] >= 'A' or str[i] <= 'F')
        if not ((condition1) or (condition2) or (condition3)):
            return False

    return True
```

**Review by Segelzwerg — COMMENTED** [review]
Fixing pylint

**Review by Segelzwerg — APPROVED** [review]

**Comment by Segelzwerg** [comment]
Sorry, I fixed a few PEP8 things to get it done.
Thanks a lot for your contribution. I hope you will come back some time.

**Pull request merged**